Broken integration test due to deprecation on master

```
pants_test/engine/legacy/test_graph_integration.py sF.F.F                [100%]

=================================== FAILURES ===================================

__________________ GraphIntegrationTest.test_existing_bundles __________________

self = <pants_test.engine.legacy.test_graph_integration.GraphIntegrationTest testMethod=test_existing_bundles>

    def test_existing_bundles(self):

      target_full = '{}:mapper'.format(self._BUNDLE_TARGET_BASE)

      pants_run = self.run_pants(['filedeps', target_full], config={

        GLOBAL_SCOPE_CONFIG_SECTION: {

          'glob_expansion_failure': 'warn',

        },

      })

      self.assert_success(pants_run)

>     self.assertNotIn("Globs", pants_run.stderr_data)

E     AssertionError: 'Globs' unexpectedly found in '/b/f/w/pants/engine/fs.py:31: DeprecationWarning: DEPRECATED: pants.util.objects.datatype will be removed in version 1.24.0.dev2.\n  Use @dataclass to declare typed named tuples instead!\n  FilesContent = Collection.of(FileContent)\n/b/f/w/pants/engine/fs.py:173: DeprecationWarning: DEPRECATED: pants.util.objects.datatype will be removed in version 1.24.0.dev2.\n  Use @dataclass to declare typed named tuples instead!\n  DirectoriesToMaterialize = Collection.of(DirectoryToMaterialize)\n/b/f/w/pants/engine/fs.py:182: DeprecationWarning: DEPRECATED: pants.util.objects.datatype will be removed in version 1.24.0.dev2.\n  Use @dataclass to declare typed named tuples instead!\n  MaterializeDirectoriesResult = Collection.of(MaterializeDirectoryResult)\n/b/f/w/pants/source/source_root.py:29: DeprecationWarning: DEPRECATED: pants.util.objects.datatype will be removed in version 1.24.0.dev2.\n  Use @dataclass to declare typed named tuples instead!\n  class AllSourceRoots(Collection.of(SourceRoot)):\n/b/f/w/pants/engine/addressable.py:15: DeprecationWarning: DEPRECATED: pants.util.objects.datatype will be removed in version 1.24.0.dev2.\n  Use @dataclass to declare typed named tuples instead!\n  Addresses = Collection.of(Address)\n/b/f/w/pants/engine/addressable.py:25: DeprecationWarning: DEPRECATED: pants.util.objects.datatype will be removed in version 1.24.0.dev2.\n  Use @dataclass to declare typed named tuples instead!\n  class BuildFileAddresses(Collection.of(BuildFileAddress)):\n/b/f/w/pants/engine/addressable.py:32: DeprecationWarning: DEPRECATED: pants.util.objects.datatype will be removed in version 1.24.0.dev2.\n  Use @dataclass to declare typed named tuples instead!\n  ProvenancedBuildFileAddresses = Collection.of(ProvenancedBuildFileAddress)\n/b/f/w/pants/engine/legacy/graph.py:381: DeprecationWarning: DEPRECATED: pants.util.objects.datatype will be removed in version 1.24.0.dev2.\n  Use @dataclass to declare typed named tuples instead!\n  class HydratedTargets(Collection.of(HydratedTarget)):\n/b/f/w/pants/engine/scheduler.py:325: DeprecationWarning: DEPRECATED: pants.util.objects.datatype will be removed in version 1.24.0.dev2.\n  Use @dataclass to declare typed named tuples instead!\n  _PathGlobsAndRootCollection = Collection.of(PathGlobsAndRoot)\n/b/f/w/pants/engine/scheduler.py:328: DeprecationWarning: DEPRECATED: pants.util.objects.datatype will be removed in version 1.24.0.dev2.\n  Use @dataclass to declare typed named tuples instead!\n  _DirectoryDigests = Collection.of(Digest)\nScrubbed PYTHONPATH=/b/f/w:/b/f/w/pytest-with-requirements.pex:/pyenv-docker-build/versions/3.6.8/lib/python36.zip:/pyenv-docker-build/versions/3.6.8/lib/python3.6:/pyenv-docker-build/versions/3.6.8/lib/python3.6/lib-dynload:/b/f/w/pex_root/install/twitter.common.log-0.3.11-py3-none-any.whl.c920dbc14b8e880aec2bb78850d101855d756952/twitter.common.log-0.3.11-py3-none-any.whl:/b/f/w/pex_root/install/docutils-0.14-py3-none-any.whl.199b17a7fbabc943293d072079258c1dd5c68e0a/docutils-0.14-py3-none-any.whl:/b/f/w/pex_root/install/python_Levenshtein-0.12.0-cp36-cp36m-linux_x86_64.whl.84166427d03452356dbc6efbc47328ff00182d55/python_Levenshtein-0.12.0-cp36-cp36m-linux_x86_64.whl:/b/f/w/pex_root/install/traceback2-1.4.0-py2.py3-none-any.whl.5e1797d06c7ad5dc1603d4320126d8c992157962/traceback2-1.4.0-py2.py3-none-any.whl:/b/f/w/pex_root/install/Pygments-2.3.1-py2.py3-none-any.whl.ee4bd72946de1ced4595abfa01a778b34f0cad86/Pygments-2.3.1-py2.py3-none-any.whl:/b/f/w/pex_root/install/chardet-3.0.4-py2.py3-none-any.whl.c4332e34e38b781695ce775973fe40663558a897/chardet-3.0.4-py2.py3-none-any.whl:/b/f/w/pex_root/install/more_itertools-7.2.0-py3-none-any.whl.f037f4802a5f30643d86c366cbf14ceb21d39f97/more_itertools-7.2.0-py3-none-any.whl:/b/f/w/pex_root/install/psutil-5.6.3-cp36-cp36m-linux_x86_64.whl.54dbf12b59be05156cf0cab2b560a8491408320e/psutil-5.6.3-cp36-cp36m-linux_x86_64.whl:/b/f/w/pex_root/install/pytest-3.6.4-py2.py3-none-any.whl.3cf4d2d91a6d84f9b62d5d5633d46f1dea9222bb/pytest-3.6.4-py2.py3-none-any.whl:/b/f/w/pex_root/install/dataclasses-0.6-py3-none-any.whl.b0182a0043d72149b99128f2badc2b43ed38de40/dataclasses-0.6-py3-none-any.whl:/b/f/w/pex_root/install/setproctitle-1.1.10-cp36-cp36m-linux_x86_64.whl.4b7f1ca0ffb03f668d5cd5acf41f15cda09e9fbf/setproctitle-1.1.10-cp36-cp36m-linux_x86_64.whl:/b/f/w/pex_root/install/certifi-2019.9.11-py2.py3-none-any.whl.9f399d4ea01e8af030fdb7731c76799123ef0839/certifi-2019.9.11-py2.py3-none-any.whl:/b/f/w/pex_root/install/pystache-0.5.3-py3-none-any.whl.0a04214ba1f2c3adf1b0141fd1b4fabe4c16109b/pystache-0.5.3-py3-none-any.whl:/b/f/w/pex_root/install/wheel-0.31.1-py2.py3-none-any.whl.9162405b4c6d47ff554018389d29281b6e4854a6/wheel-0.31.1-py2.py3-none-any.whl:/b/f/w/pex_root/install/packaging-16.8-py2.py3-none-any.whl.e40536ef84ff020576b969f64434fa248c714237/packaging-16.8-py2.py3-none-any.whl:/b/f/w/pex_root/install/twitter.common.lang-0.3.11-py3-none-any.whl.753335ecc81fb9377df5ad090881a9a3cfaf6aa9/twitter.common.lang-0.3.11-py3-none-any.whl:/b/f/w/pex_root/install/six-1.12.0-py2.py3-none-any.whl.56417fecea1347c6c047607ab9996cc97aec99de/six-1.12.0-py2.py3-none-any.whl:/b/f/w/pex_root/install/twitter.common.collections-0.3.11-py3-none-any.whl.b3eeeb6a640fca93f1e271a2dd9a8a81fb47c030/twitter.common.collections-0.3.11-py3-none-any.whl:/b/f/w/pex_root/install/pluggy-0.7.1-py2.py3-none-any.whl.742f368de18a3c89563e6d1a269f2b5b7f336d2d/pluggy-0.7.1-py2.py3-none-any.whl:/b/f/w/pex_root/install/requests-2.22.0-py2.py3-none-any.whl.bb3699cc576016729d42f9bf4eccfdbcb292ec50/requests-2.22.0-py2.py3-none-any.whl:/b/f/w/pex_root/install/cffi-1.12.3-cp36-cp36m-manylinux1_x86_64.whl.73be7de09a609474b1efc2d12432becbbd0f0de4/cffi-1.12.3-cp36-cp36m-manylinux1_x86_64.whl:/b/f/w/pex_root/install/twitter.common.options-0.3.11-py3-none-any.whl.136eca7ec58c4a151b2d5b6a71d631ed1b3658b8/twitter.common.options-0.3.11-py3-none-any.whl:/b/f/w/pex_root/install/pyOpenSSL-17.3.0-py2.py3-none-any.whl.f542e1c53dc88676d30ec1b1da93018ff42cb6a9/pyOpenSSL-17.3.0-py2.py3-none-any.whl:/b/f/w/pex_root/install/linecache2-1.0.0-py2.py3-none-any.whl.fd2cf72046aaaf68615409af851242d5c32b321f/linecache2-1.0.0-py2.py3-none-any.whl:/b/f/w/pex_root/install/idna-2.8-py2.py3-none-any.whl.645e82ab4157606d15a8766060dd690deebcc4f2/idna-2.8-py2.py3-none-any.whl:/b/f/w/pex_root/install/asttokens-1.1.13-py2.py3-none-any.whl.00fa8ee883d8fd945420c0f59d79e2b8387d013b/asttokens-1.1.13-py2.py3-none-any.whl:/b/f/w/pex_root/install/twitter.common.dirutil-0.3.11-py3-none-any.whl.27719e6d70e5792a4eaa4b057e7ad31a39bceddd/twitter.common.dirutil-0.3.11-py3-none-any.whl:/b/f/w/pex_root/install/ply-3.11-py2.py3-none-any.whl.ced24d23e3ca145d26f6ac130df2de53c3e35188/ply-3.11-py2.py3-none-any.whl:/b/f/w/pex_root/install/py-1.8.0-py2.py3-none-any.whl.3402fa58a6c20477b2e9efe86f4f43a119cd771a/py-1.8.0-py2.py3-none-any.whl:/b/f/w/pex_root/install/pathspec-0.5.9-py3-none-any.whl.8385d9631afea5f89b0b84977cd9d811d214d314/pathspec-0.5.9-py3-none-any.whl:/b/f/w/pex_root/install/urllib3-1.25.6-py2.py3-none-any.whl.be3132b0776d8d906e57978f1591774617a1d4fb/urllib3-1.25.6-py2.py3-none-any.whl:/b/f/w/pex_root/install/pytest_timeout-1.2.1-py2.py3-none-any.whl.4581b560be35b67389220084292ed6b31c056bc3/pytest_timeout-1.2.1-py2.py3-none-any.whl:/b/f/w/pex_root/install/PyYAML-5.1.2-cp36-cp36m-linux_x86_64.whl.9cf42c8cdcdf7383fe851f85465a864fa2f34348/PyYAML-5.1.2-cp36-cp36m-linux_x86_64.whl:/b/f/w/pex_root/install/pywatchman-1.4.1-cp36-cp36m-linux_x86_64.whl.fff5da8bcc13fc32a429b12491a3a71f44257e42/pywatchman-1.4.1-cp36-cp36m-linux_x86_64.whl:/b/f/w/pex_root/install/attrs-19.3.0-py2.py3-none-any.whl.01c47aecb2cee3cc26140dc70eb5d5066d174785/attrs-19.3.0-py2.py3-none-any.whl:/b/f/w/pex_root/install/twitter.common.confluence-0.3.11-py3-none-any.whl.db65f7c09e90802248102d3fe380f367d26b066e/twitter.common.confluence-0.3.11-py3-none-any.whl:/b/f/w/pex_root/install/beautifulsoup4-4.6.3-py3-none-any.whl.f69b3796d967a649e5bef9622259dadb21e19d59/beautifulsoup4-4.6.3-py3-none-any.whl:/b/f/w/pex_root/install/coverage-4.5.4-cp36-cp36m-manylinux1_x86_64.whl.53442ad0427aece4db45460c3b9db6b525598f8f/coverage-4.5.4-cp36-cp36m-manylinux1_x86_64.whl:/b/f/w/pex_root/install/fasteners-0.14.1-py2.py3-none-any.whl.99944bb9e5cc9cc4f6144bc3ee6173a78219b35a/fasteners-0.14.1-py2.py3-none-any.whl:/b/f/w/pex_root/install/pytest_cov-2.4.0-py2.py3-none-any.whl.18334e5c5b4fe92fac23389f65776f52e363bbd5/pytest_cov-2.4.0-py2.py3-none-any.whl:/b/f/w/pex_root/install/setuptools-40.6.3-py2.py3-none-any.whl.4a47c8d55bbea45ef1029b493d822f7c1d55a196/setuptools-40.6.3-py2.py3-none-any.whl:/b/f/w/pex_root/install/unittest2-1.1.0-py2.py3-none-any.whl.825429223a8ef4a056925c48c01a7f7a387cb1fd/unittest2-1.1.0-py2.py3-none-any.whl:/b/f/w/pex_root/install/pex-1.6.12-py2.py3-none-any.whl.d8a368099663ef2d23c79e3403d86115039740d7/pex-1.6.12-py2.py3-none-any.whl:/b/f/w/pex_root/install/atomicwrites-1.3.0-py2.py3-none-any.whl.2bce967cb4fb97fa3931f45e5b4decdf527bf276/atomicwrites-1.3.0-py2.py3-none-any.whl:/b/f/w/pex_root/install/thriftpy2-0.4.7-cp36-cp36m-linux_x86_64.whl.21f6416b61ab690d8454b1a6cf8c6cfee0a61d8f/thriftpy2-0.4.7-cp36-cp36m-linux_x86_64.whl:/b/f/w/pex_root/install/pycparser-2.19-py2.py3-none-any.whl.53078ed0a121a74ad0e5d5d8534cec1c8f49234e/pycparser-2.19-py2.py3-none-any.whl:/b/f/w/pex_root/install/pycodestyle-2.4.0-py2.py3-none-any.whl.0a958eee162effa42dbebab9ec410ad2e5c7fa7b/pycodestyle-2.4.0-py2.py3-none-any.whl:/b/f/w/pex_root/install/monotonic-1.5-py2.py3-none-any.whl.e6c7e442a796dc3284dafc29d5bee78b3d3b527a/monotonic-1.5-py2.py3-none-any.whl:/b/f/w/pex_root/install/pyflakes-2.1.1-py2.py3-none-any.whl.728507c2086c87982c8c1c24ff46b1c2e6c46574/pyflakes-2.1.1-py2.py3-none-any.whl:/b/f/w/pex_root/install/cryptography-2.8-cp36-cp36m-linux_x86_64.whl.fb86a63dbcd99734250a2790f7b1f4dc63133b91/cryptography-2.8-cp36-cp36m-linux_x86_64.whl:/b/f/w/pex_root/install/contextlib2-0.5.5-py2.py3-none-any.whl.edad6efe1aebf29bea2d8a9e8dab93fbbc2555bc/contextlib2-0.5.5-py2.py3-none-any.whl:/b/f/w/pex_root/install/argparse-1.4.0-py2.py3-none-any.whl.dd8addeaf2a7bb1139fae1ac955d19f7c60bd9b2/argparse-1.4.0-py2.py3-none-any.whl:/b/f/w/pex_root/install/ansicolors-1.0.2-py3-none-any.whl.0088834e21b2c8f3df9b82794f4dca1c75af4f12/ansicolors-1.0.2-py3-none-any.whl:/b/f/w/pex_root/install/www_authenticate-0.9.2-py3-none-any.whl.a7cb2145b3e03bc83c68b6a3bd8f7d26627568b4/www_authenticate-0.9.2-py3-none-any.whl:/b/f/w/pex_root/install/typing_extensions-3.7.4-py3-none-any.whl.1edcbe8619e77859668b3f056ad84bcfdf2e0fcd/typing_extensions-3.7.4-py3-none-any.whl:/b/f/w/pex_root/install/Markdown-2.1.1-py3-none-any.whl.6456b7a4f7d5e30ef5882665fd6c267d5996a993/Markdown-2.1.1-py3-none-any.whl:/b/f/w/pex_root/install/py_zipkin-0.18.4-py2.py3-none-any.whl.cf7543ab85c6b8493274c465aa3e70d8bedd4118/py_zipkin-0.18.4-py2.py3-none-any.whl:/b/f/w/pex_root/install/pyparsing-2.4.2-py2.py3-none-any.whl.50391d7edbea580d5d233c392696aef4d9056311/pyparsing-2.4.2-py2.py3-none-any.whl:/b/f/w/pytest-with-requirements.pex/.bootstrap from the environment.\n/b/f/w/pants/backend/project_info/rules/source_file_validator.py:96: DeprecationWarning: DEPRECATED: pants.util.objects.datatype will be removed in version 1.24.0.dev2.\n  Use @dataclass to declare typed named tuples instead!\n  RegexMatchResults = Collection.of(RegexMatchResult)\n\n'

pants_test/engine/legacy/test_graph_integration.py:143: AssertionError
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -28,7 +28,6 @@ env:
   - AWS_ACCESS_KEY_ID__TO_BE_REEXPORTED_ON_DEPLOYS=AKIAV6A6G7RQWPRUWIXR
   - secure: hFVAQGLVkexzTd3f9NF+JoG1dE+CPICKqOcdvQYv8+YB2rwwqu0/J6MnqKUZSmec4AM4ZvyPUBIHnSw8aMJysYs+GZ6iG/8ZRRmdbmo2WBPbSZ+ThRZxx/F6AjmovUmf8Zt366ZAZXpc9NHKREkTUGl6UL7FFe9+ouVnb90asdw=
   - RUST_BACKTRACE="all"
-  - PYTHONWARNINGS=ignore::DeprecationWarning
 matrix:
   include:
   - addons:

--- a/build-support/bin/generate_travis_yml.py
+++ b/build-support/bin/generate_travis_yml.py
@@ -70,7 +70,6 @@ GLOBAL_ENV_VARS = [
   # https://docs.travis-ci.com/user/environment-variables#defining-encrypted-variables-in-travisyml.
   {"secure": "hFVAQGLVkexzTd3f9NF+JoG1dE+CPICKqOcdvQYv8+YB2rwwqu0/J6MnqKUZSmec4AM4ZvyPUBIHnSw8aMJysYs+GZ6iG/8ZRRmdbmo2WBPbSZ+ThRZxx/F6AjmovUmf8Zt366ZAZXpc9NHKREkTUGl6UL7FFe9+ouVnb90asdw="},
   'RUST_BACKTRACE="all"',
-  'PYTHONWARNINGS=ignore::DeprecationWarning',
 ]
 
 # ----------------------------------------------------------------------

--- a/src/python/pants/base/deprecated.py
+++ b/src/python/pants/base/deprecated.py
@@ -234,8 +234,7 @@ def deprecated(
   removal_version: str,
   hint_message: Optional[str] = None,
   subject: Optional[str] = None,
-  ensure_stderr: bool = False,
-  stacklevel: int = 3
+  ensure_stderr: bool = False
 ):
   """Marks a function or method as deprecated.
 
@@ -269,8 +268,7 @@ def deprecated(
     @wraps(func)
     def wrapper(*args, **kwargs):
       warn_or_error(removal_version, subject or func_full_name, hint_message,
-                    ensure_stderr=ensure_stderr,
-                    stacklevel=stacklevel)
+                    ensure_stderr=ensure_stderr)
       return func(*args, **kwargs)
     return wrapper
   return decorator

--- a/src/python/pants/bin/pants_loader.py
+++ b/src/python/pants/bin/pants_loader.py
@@ -28,8 +28,7 @@ class PantsLoader:
     #
     # However, we do turn off deprecation warnings for libraries that Pants uses for which we do
     # not have a fixed upstream version, typically because the library is no longer maintained.
-    if not os.environ.get('PYTHONWARNINGS'):
-      warnings.simplefilter('default', category=DeprecationWarning)
+    warnings.simplefilter('default', category=DeprecationWarning)
     # TODO: Eric-Arellano has emailed the author to see if he is willing to accept a PR fixing the
     # deprecation warnings and to release the fix. If he says yes, remove this once fixed.
     warnings.filterwarnings('ignore', category=DeprecationWarning, module="ansicolors")
@@ -48,7 +47,7 @@ class PantsLoader:
         Fix it by setting the LC_* and LANG environment settings. Example:
           LC_ALL=en_US.UTF-8
           LANG=en_US.UTF-8
-        Or, bypass it by setting the below environment variable.
+        Or, bypass it by setting the below environment variable. 
           {}=1
         Note: we cannot guarantee consistent behavior with this bypass enabled.
         """.format(encoding, cls.ENCODING_IGNORE_ENV_VAR)

--- a/src/python/pants/engine/BUILD
+++ b/src/python/pants/engine/BUILD
@@ -164,7 +164,6 @@ python_library(
   sources=['rules.py'],
   dependencies=[
     '3rdparty/python:asttokens',
-    '3rdparty/python:dataclasses',
     '3rdparty/python/twitter/commons:twitter.common.collections',
     ':goal',
     ':selectors',

--- a/src/python/pants/engine/README.md
+++ b/src/python/pants/engine/README.md
@@ -28,7 +28,7 @@ produce the requested value.
 ### Products and Params
 
 The engine executes your `@rule`s in order to (recursively) compute a `Product` of the requested
-type for a set of `Param`s. This recursive type search leads to a loosely coupled (and yet still
+type for a set of `Param`s. This recursive type search leads to a loosely coupled (and yet still 
 statically checked) form of dependency injection.
 
 When an `@rule` runs, it requires a set of `Param`s that the engine has determined are needed to
@@ -76,72 +76,52 @@ In cases where this search detects any ambiguity (generally because there are tw
 that can provide the same product with the same number of parameters), rule graph compilation will
 fail with a useful error message.
 
-### How to Declare Engine `Param` Types
+### Datatypes
 
 In practical use, builtin types like `str` or `int` do not provide enough information to
-disambiguate between various types of data in `@rule` signatures, so declaring small, unique classes
-to encapsulate different states can help to make `@rule` sets with complex control flow between rules
-more declarative and self-documenting.
-
-#### Requirements for an Engine `Param`
-
-To use an instance of some class `C` as a `Param` to the engine:
-1. instances of `C` must be immutable,
-2. `__hash__` and `__eq__` must be implemented such that when constructing two separate instances of `C` with the same argument list `...`, `C(...) == C(...)` and `hash(C(...)) == hash(C(...))`.
-  - This can be ignored for singleton `Param`s.
-
-#### Benefits of using `@dataclass` for a `Param`
-[Python 3 `@dataclass`es](https://docs.python.org/3/library/dataclasses.html) satisfy the above requirements for engine `Param`s. Using `@dataclass` to declare engine `Param` types also provides additional benefits compared to alternatives:
-1. a compact and high-performance representation which can be stably shared across FFI boundaries,
-2. static type checking via [the `dataclasses` mypy plugin](https://github.com/python/mypy/blob/master/mypy/plugins/dataclasses.py),
-3. a concise, standard, and Pythonic way to declare classes.
-
-
-#### Example Usage of `@dataclass` for Engine `Param`s
-
-*Note that the [3rdparty `dataclasses` library](https://github.com/ericvsmith/dataclasses) must be in your BUILD file and `import`ed if you're running Pants with Python 3.6!*
+disambiguate between various types of data in `@rule` signatures, so declaring small `datatype` 
+definitions to provide a unique and descriptive type is highly recommended:
 
 ```python
-from dataclasses import dataclass
-
-from pants.util.meta import frozen_after_init
-
-# Pants requires that engine Params have a stable hash. This can be accomplished with the
-# `frozen=True` argument set in the `@dataclass` decorator.
-@dataclass(frozen=True)
-class FormattedInt:
-  content: str
-
-# `@frozen_after_init` can also be used with `unsafe_hash=True` to create engine Params which can
-# modify their fields within the `__init__()` method:
-@frozen_after_init
-@dataclass(unsafe_hash=True)
-class ValidatedInt:
-  x: int
-
-  def __init__(self, x):
-    self.x = x
+class FormattedInt(datatype(['content'])): pass
 
 @rule
-def int_to_str(value: ValidatedInt) -> FormattedInt:
-  return FormattedInt('{}'.format(value.x))
+def int_to_str(value: int) -> FormattedInt:
+  return FormattedInt('{}'.format(value))
 
-# Instances can be constructed and accessed the same way as for normal `@dataclass`es.
+# Field values can be specified with positional and/or keyword arguments in the constructor:
 x = FormattedInt('a string')
-assert x == FormattedInt(content='a string')
-print(x.content)                # 'a string'
+x = FormattedInt(content='a string')
 
-# `@dataclass` objects can be easily inspected:
-print(x)                        # 'FormattedInt(content="a string")'
+# Field values can be accessed after construction by name or index:
+print(x.content)    # 'a string'
+print(x[0])         # 'a string'
 
-# All mypy types, including parameterized types, may be used in field definitions.
-@dataclass(frozen=True)
-class TypedDatatype:
-  """Example of a datatype with a more complex field type constraint."""
-  field_name: Union[str, int]
-
-print(TypedDatatype("huh")) # 'TypedDatatype(field_name=huh)'
+# datatype objects can be easily inspected:
+print(x)            # 'FormattedInt(content=a string)'
 ```
+
+#### Types of Fields
+
+`datatype()` accepts a list of *field declarations*, and returns a type which can be subclassed. A
+*field declaration* can just be a string (e.g. `'field_name'`), which is then used as the field
+name, as with `FormattedInt` above. A field can also be declared with a tuple of two elements: the
+field name string, and a `TypeConstraint` for the field (e.g. `('field_name',
+Exactly(FieldType))`). The bare type name (e.g. `FieldType`) can also be used as a shorthand for
+`Exactly(FieldType)`. If the tuple form is used, the constructor will create your object, then raise
+an error if the field value does not satisfy the type constraint.
+
+``` python
+class TypedDatatype(datatype([('field_name', Exactly(str, int))])):
+  """Example of a datatype with a more complex field type constraint."""
+```
+
+Assigning a specific type to a field can be somewhat unidiomatic in Python, and may be unexpected or
+unnatural to use. However, regardless of whether the object is created directly with type-checked
+fields or whether it's produced from a set of rules by the engine's dependency injection, it is
+extremely useful to formalize the assumptions made about the value of an object into a specific
+type, even if the type just wraps a single field. The `datatype()` function makes it simple and
+efficient to apply that strategy.
 
 ### Gets and RootRules
 

--- a/src/python/pants/engine/rules.py
+++ b/src/python/pants/engine/rules.py
@@ -2,7 +2,6 @@
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
 import ast
-import dataclasses
 import inspect
 import itertools
 import logging
@@ -13,7 +12,7 @@ from collections import OrderedDict
 from collections.abc import Iterable
 from dataclasses import dataclass
 from textwrap import dedent
-from typing import Any, Callable, Dict, Optional, Set, Tuple, Type, cast
+from typing import Any, Callable, Dict, Optional, Tuple, Type, cast
 
 import asttokens
 from twitter.common.collections import OrderedSet
@@ -466,19 +465,6 @@ class Rule(ABC):
     """A tuple of Optionable classes that are known to be necessary to run this rule."""
     return ()
 
-  @classmethod
-  def _validate_type_field(cls, type_obj, description):
-    if not isinstance(type_obj, type):
-      raise TypeError(f'{description} provided to @rules must be types! Was: {type_obj}.')
-    if dataclasses.is_dataclass(type_obj):
-      if not (type_obj.__dataclass_params__.frozen or
-              getattr(type_obj, frozen_after_init.sentinel_attr, False)):
-        raise TypeError(
-          f'{description} {type_obj} is a dataclass declared without `frozen=True`, or without '
-          'both `unsafe_hash=True` and the `@frozen_after_init` decorator! '
-          'The engine requires that fields in params are immutable for stable hashing!')
-    return type_obj
-
 
 @frozen_after_init
 @dataclass(unsafe_hash=True)
@@ -507,16 +493,8 @@ class TaskRule(Rule):
     dependency_optionables: Optional[Tuple] = None,
     cacheable: bool = True,
   ):
-    self._output_type = self._validate_type_field(output_type, '@rule output type')
-    self.input_selectors = tuple(
-      self._validate_type_field(t, '@rule input selector')
-      for t in input_selectors
-    )
-    for g in input_gets:
-      product_type = g.product
-      subject_type = g.subject_declared_type
-      self._validate_type_field(product_type, 'Get product type')
-      self._validate_type_field(subject_type, 'Get subject type')
+    self._output_type = output_type
+    self.input_selectors = input_selectors
     self.input_gets = input_gets
     self.func = func
     self._dependency_rules = dependency_rules or ()
@@ -556,7 +534,7 @@ class RootRule(Rule):
   _output_type: Type
 
   def __init__(self, output_type: Type) -> None:
-    self._output_type = self._validate_type_field(output_type, 'RootRule declared type')
+    self._output_type = output_type
 
   @property
   def output_type(self):
@@ -571,12 +549,13 @@ class RootRule(Rule):
     return tuple()
 
 
+# TODO: add typechecking here -- use dicts for `union_rules`.
 @dataclass(frozen=True)
 class RuleIndex:
   """Holds a normalized index of Rules used to instantiate Nodes."""
-  rules: Dict[Any, Any]
-  roots: Set[Any]
-  union_rules: Dict[Any, Any]
+  rules: Any
+  roots: Any
+  union_rules: Any
 
   @classmethod
   def create(cls, rule_entries, union_rules=None):

--- a/src/python/pants/util/BUILD
+++ b/src/python/pants/util/BUILD
@@ -98,8 +98,6 @@ python_library(
   sources = ['objects.py'],
   dependencies = [
     '3rdparty/python/twitter/commons:twitter.common.collections',
-    'src/python/pants/base:deprecated',
-    ':collections',
     ':strutil',
     ':memo',
     ':meta',

--- a/src/python/pants/util/meta.py
+++ b/src/python/pants/util/meta.py
@@ -137,8 +137,4 @@ def frozen_after_init(cls: Type[T]) -> Type[T]:
 
   cls.__init__ = new_init  # type: ignore
   cls.__setattr__ = new_setattr  # type: ignore
-  setattr(cls, frozen_after_init.sentinel_attr, True)
-
   return cls
-
-frozen_after_init.sentinel_attr = '_frozen_after_init'

--- a/src/python/pants/util/meta.py
+++ b/src/python/pants/util/meta.py
@@ -137,8 +137,8 @@ def frozen_after_init(cls: Type[T]) -> Type[T]:
 
   cls.__init__ = new_init  # type: ignore
   cls.__setattr__ = new_setattr  # type: ignore
-  setattr(cls, frozen_after_init.sentinel_attr, True) # type: ignore
+  setattr(cls, frozen_after_init.sentinel_attr, True)
 
   return cls
 
-frozen_after_init.sentinel_attr = '_frozen_after_init' # type: ignore
+frozen_after_init.sentinel_attr = '_frozen_after_init'

--- a/src/python/pants/util/objects.py
+++ b/src/python/pants/util/objects.py
@@ -5,7 +5,6 @@ from abc import ABC, abstractmethod
 from collections import OrderedDict, namedtuple
 from collections.abc import Iterable
 
-from pants.base.deprecated import deprecated
 from pants.util.memo import memoized_classproperty
 from pants.util.meta import classproperty
 from pants.util.strutil import pluralize
@@ -44,8 +43,12 @@ class DatatypeMixin(ABC):
     """
     return cls.type_check_error_type(cls.__name__, msg, *args, **kwargs)
 
+  @abstractmethod
+  def copy(self, **kwargs):
+    """Return a new object of the same type, replacing specified fields with new values"""
 
-@deprecated('1.24.0.dev2', hint_message='Use @dataclass to declare typed named tuples instead!', stacklevel=5)
+
+# TODO(#7074): Migrate to python 3 dataclasses!
 def datatype(field_decls, superclass_name=None, **kwargs):
   """A wrapper for `namedtuple` that accounts for the type of the object in equality.
 
@@ -398,8 +401,14 @@ class TypedCollection(TypeConstraint):
   def __init__(self, constraint):
     """Create a `TypeConstraint` which validates each member of a collection with `constraint`.
 
-    :param TypeConstraint constraint: the `TypeConstraint` to apply to each element.
+    :param TypeOnlyConstraint constraint: the `TypeConstraint` to apply to each element. This is
+                                          currently required to be a `TypeOnlyConstraint` to avoid
+                                          complex prototypal type relationships.
     """
+
+    if not isinstance(constraint, TypeOnlyConstraint):
+      raise TypeError("constraint for collection must be a {}! was: {}"
+                      .format(TypeOnlyConstraint.__name__, constraint))
 
     description = '{}({})'.format(type(self).__name__, constraint)
 

--- a/tests/python/pants_test/engine/BUILD
+++ b/tests/python/pants_test/engine/BUILD
@@ -154,7 +154,7 @@ python_tests(
     'src/python/pants/util:dirutil',
     'tests/python/pants_test/engine/examples:mapper_test',
     'tests/python/pants_test/engine/examples:parsers',
-  ],
+  ]
 )
 
 python_tests(

--- a/tests/python/pants_test/engine/test_scheduler.py
+++ b/tests/python/pants_test/engine/test_scheduler.py
@@ -13,7 +13,6 @@ from pants.engine.native import Native
 from pants.engine.rules import RootRule, UnionRule, rule, union
 from pants.engine.scheduler import ExecutionError, SchedulerSession
 from pants.engine.selectors import Get, Params
-from pants.util.meta import frozen_after_init
 from pants_test.engine.util import assert_equal_with_printing, remove_locations_from_traceback
 from pants_test.test_base import TestBase
 
@@ -61,43 +60,6 @@ class D:
 def transitive_coroutine_rule(c: C) -> D:
   b = yield Get(B, C, c)
   yield D(b)
-
-
-@dataclass
-class NonFrozenDataclass:
-  x: int
-
-
-@frozen_after_init
-@dataclass(unsafe_hash=True)
-class FrozenAfterInit:
-  x: int
-
-  def __init__(self, x):
-    # This is an example of how you can assign within __init__() with @frozen_after_init. This
-    # particular example is not intended to be super useful.
-    self.x = x + 1
-
-
-@rule
-def use_frozen_after_init_object(x: FrozenAfterInit) -> int:
-  return x.x
-
-
-@dataclass(frozen=True)
-class FrozenFieldsDataclass:
-  x: int
-  y: str
-
-
-@dataclass(frozen=True)
-class ResultDataclass:
-  something: str
-
-
-@rule
-def dataclass_rule(obj: FrozenFieldsDataclass) -> ResultDataclass:
-  return ResultDataclass(something=f'x={obj.x}, y={obj.y}')
 
 
 @union
@@ -217,10 +179,6 @@ class SchedulerTest(TestBase):
       consumes_a_and_b,
       transitive_b_c,
       transitive_coroutine_rule,
-      dataclass_rule,
-      RootRule(FrozenAfterInit),
-      use_frozen_after_init_object,
-      RootRule(FrozenFieldsDataclass),
       RootRule(UnionWrapper),
       UnionRule(UnionBase, UnionA),
       UnionRule(UnionWithNonMemberErrorMsg, UnionX),
@@ -285,17 +243,6 @@ Type A is not a member of the UnionBase @union
     expected_msg = "specific error message for UnionA instance"
     with self._assert_execution_error(expected_msg):
       self.scheduler.product_request(UnionX, [Params(UnionWrapper(UnionA()))])
-
-  def test_dataclass_products_rule(self):
-    result, = self.scheduler.product_request(
-      ResultDataclass,
-      [Params(FrozenFieldsDataclass(3, "test string"))])
-    self.assertEquals(result.something, 'x=3, y=test string')
-
-    result, = self.scheduler.product_request(
-      int,
-      [Params(FrozenAfterInit(x=3))])
-    self.assertEquals(result, 4)
 
 
 class SchedulerWithNestedRaiseTest(TestBase):
@@ -410,23 +357,3 @@ Exception: WithDeps(Inner(InnerEntry { params: {TypeCheckFailWrapper}, rule: Tas
                                raise Exception(f'An exception for {type(x).__name__}')
                            Exception: An exception for B''').lstrip() + '\n\n',  # Traces include two empty lines after.
                                trace)
-
-
-class RuleIndexingErrorTest(TestBase):
-
-  def test_non_frozen_dataclass_error(self):
-    with self.assertRaisesWithMessage(TypeError, dedent("""\
-      RootRule declared type <class 'pants_test.engine.test_scheduler.NonFrozenDataclass'> is a dataclass declared without `frozen=True`, or without both `unsafe_hash=True` and the `@frozen_after_init` decorator! The engine requires that fields in params are immutable for stable hashing!""")):
-      RootRule(NonFrozenDataclass)
-
-    with self.assertRaisesWithMessage(TypeError, dedent("""\
-      @rule input selector <class 'pants_test.engine.test_scheduler.NonFrozenDataclass'> is a dataclass declared without `frozen=True`, or without both `unsafe_hash=True` and the `@frozen_after_init` decorator! The engine requires that fields in params are immutable for stable hashing!""")):
-      @rule
-      def f(x: NonFrozenDataclass) -> int:
-        return 3
-
-    with self.assertRaisesWithMessage(TypeError, dedent("""\
-      @rule output type <class 'pants_test.engine.test_scheduler.NonFrozenDataclass'> is a dataclass declared without `frozen=True`, or without both `unsafe_hash=True` and the `@frozen_after_init` decorator! The engine requires that fields in params are immutable for stable hashing!""")):
-      @rule
-      def f(x: int) -> NonFrozenDataclass:
-        return NonFrozenDataclass(x=x)

--- a/tests/python/pants_test/engine/test_scheduler.py
+++ b/tests/python/pants_test/engine/test_scheduler.py
@@ -428,5 +428,5 @@ class RuleIndexingErrorTest(TestBase):
     with self.assertRaisesWithMessage(TypeError, dedent("""\
       @rule output type <class 'pants_test.engine.test_scheduler.NonFrozenDataclass'> is a dataclass declared without `frozen=True`, or without both `unsafe_hash=True` and the `@frozen_after_init` decorator! The engine requires that fields in params are immutable for stable hashing!""")):
       @rule
-      def g(x: int) -> NonFrozenDataclass:
+      def f(x: int) -> NonFrozenDataclass:
         return NonFrozenDataclass(x=x)

--- a/tests/python/pants_test/pants_run_integration_test.py
+++ b/tests/python/pants_test/pants_run_integration_test.py
@@ -213,8 +213,6 @@ class PantsRunIntegrationTest(unittest.TestCase):
         # Ensure that the underlying ./pants invocation doesn't run from sources
         # (and therefore bootstrap) if we don't want it to.
         'RUN_PANTS_FROM_PEX',
-        # Used to ignore `DeprecationWarning`s in CI.
-        'PYTHONWARNINGS',
       ]
 
   def setUp(self):

--- a/tests/python/pants_test/util/test_objects.py
+++ b/tests/python/pants_test/util/test_objects.py
@@ -190,16 +190,12 @@ class TypedCollectionTest(TypeConstraintTestBase):
     self.assertTrue(collection_constraint.satisfied_by([self.B(), self.C(), self.BPrime()]))
     self.assertFalse(collection_constraint.satisfied_by([self.B(), self.A()]))
 
-  def test_complex_sub_constraint(self):
+  def test_no_complex_sub_constraint(self):
     sub_collection = TypedCollection(Exactly(self.A))
-    nested_collection = TypedCollection(sub_collection)
-    self.assertTrue(nested_collection.satisfied_by(()))
-    self.assertTrue(nested_collection.satisfied_by([]))
-
-    self.assertTrue(nested_collection.satisfied_by([[]]))
-    self.assertTrue(nested_collection.satisfied_by([[self.A()]]))
-
-    self.assertFalse(nested_collection.satisfied_by([[self.B()]]))
+    with self.assertRaisesWithMessage(
+        TypeError,
+        "constraint for collection must be a TypeOnlyConstraint! was: {}".format(sub_collection)):
+      TypedCollection(sub_collection)
 
   def test_validate(self):
     collection_exactly_a_or_b = TypedCollection(Exactly(self.A, self.B))

